### PR TITLE
Kadanes_algorithm

### PR DESCRIPTION
**Pull request by Raj-Parekh24**
### **Describe your change:**

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [ ] This pull request is all my own work -- I have not plagiarized.
* [ ] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: maths/Kadane's_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            max_till_element = 0
+    
+    return max_sum
+
+
+
+if __name__ == "__main__":
+    try:
+        print("Enter the element of array with spaces  :- ")
+        arr = list(map(int, input().split()))
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py:65]:
```suggestion
        arr = [int(x) for x in input().split()]
```

File: maths/Kadane's_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        print(kadanes_implementation(arr))
+    except ValueError:
+        print("Please,enter decimal values")
+
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py:69]:
```suggestion
```

File: maths/Kadane's_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+     Eg :- 
+     arr = [2,3,-9,8,-2]
+       initially the value of max_sum = 0 and max_till_element is 0
+       than when max_sum less than max_till particular element it will assign that value to max_sum
+       and when value of max_till_sum is less than 0 it will assign 0 to i
+       and after whole process it will return the value
+
+       So th output for above arr is :- 8
+    """
+    if(negative_exist(arr)<0):
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py:45]:
```suggestion
    if negative_exist(arr) < 0:
```

File: maths/Kadane's_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+     else it will return the value return by negative_exist function
+     
+     Eg :- 
+     arr = [2,3,-9,8,-2]
+       initially the value of max_sum = 0 and max_till_element is 0
+       than when max_sum less than max_till particular element it will assign that value to max_sum
+       and when value of max_till_sum is less than 0 it will assign 0 to i
+       and after whole process it will return the value
+
+       So th output for above arr is :- 8
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py:43]:
```suggestion
       So the output for above arr is :- 8
```
Please replace this comment with a doctest.

File: maths/Kadane's_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    max=arr[0]
+    for i in arr:
+        if(i>=0):
+            return 0;
+        elif(max<=i):
+            max=i
+    return max 
+
+
+def kadanes_implementation(arr):
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py:29]: Python type hints as discussed in CONTRIBUTING.md?

File: maths/Kadane's_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+     eg :- 
+      arr = [-2,-8,-9]
+      so max element is negative i.e. -2
+      so answer is -2
+
+      and if max value is positive it will return 0
+      and then apply the kadane's algorithm
+      eg :- 
+      arr = [2,8,-9]
+      positive number exist so it will return 0
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py:18]: Please add doctests.

File: maths/Kadane's_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+"""
+    Kadane's algorithm to get maximum subarray sum
+    Please enter the decimal values only and use spaces as seperator
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py:3]: Please add URL to a website that explains the algorithm and why it would be useful.

**Comment by cclauss** [comment]
Codespell test is failing...  Travis test too.
```
./maths/Kadane's_algorithm.py:3: seperator  ==> separator
./maths/Kadane's_algorithm.py:8: wether  ==> weather, whether
```

File: maths/Kadane's_algorithm.py (hunk 7 above)
Comment (Raj-Parekh24) [reply maths/Kadane's_algorithm.py]: Thanks for the suggestions
As I am new to open source contribution this help is meant a lot to me.

File: maths/Kadane's_algorithm.py [hunk 8]
```diff
@@ -62,8 +77,8 @@ def kadanes_implementation(arr):
 if __name__ == "__main__":
     try:
         print("Enter the element of array with spaces  :- ")
-        arr = list(map(int, input().split()))
-        print(kadanes_implementation(arr))
+        arr = [int(x) for x in input().split()]
+        print("Maximum subarray sum of ",arr,"is :-",kadanes_implementation(arr))
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py]:
```suggestion
        print(f"Maximum subarray sum of {arr} is {kadanes_implementation(arr)}")

```

File: maths/Kadane's_algorithm.py [hunk 9]
```diff
@@ -1,11 +1,26 @@
...
+Enter the element of array with spaces  :-
+>>> 2 3 -9 8 -2 
+Maximum subarray sum of  [2, 3, -9, 8, -2] is :- 8
+
+
+Enter the element of array with spaces  :-
+>>> -9 -8 -7 -6 -2
+Maximum subarray sum of  [-9, -8, -7, -6, -2] is :- -2
+
 """
```
Comment (cclauss) [inline maths/Kadane's_algorithm.py]: Doctest will not run if it is not in a docstring.  The second multiline string is not a docstring but the first is.  Please move these tests into the _docstrings of the functions_.  Please run `python3 -m doctest -v Kadane's_algorithm.py` on your own computer to make sure that they run and pass.

**Comment by Raj-Parekh24** [comment]
Changes suggested by you has been done.

Please go through the code and suggest me any further changes.

Thanks for the support..

**Comment by Raj-Parekh24** [comment]
Changes suggested by you has been implemented.

Please go through it.

File: maths/kadanes_algorithm.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
+"""
+    Link for the explanation of this algorithm :- 
+    https://medium.com/@rsinghal757/kadanes-algorithm-dynamic-programming-how-and-why-does-it-work-3fd8849ed73d
+    Kadane's algorithm to get maximum subarray sum
+    Please enter the decimal values only and use spaces as separator
+"""
+def negative_exist(arr):
```
Comment (cclauss) [inline maths/kadanes_algorithm.py:7]:
```suggestion
test_data = ([-2, -8, -9], [2, 8, 9], [-1, 0, 1], [0, 0], [])


def negative_exist(arr: list) -> int:
```
Different lists: all negative, all positive, mixed, all zero, empty list.

File: maths/kadanes_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    Link for the explanation of this algorithm :- 
+    https://medium.com/@rsinghal757/kadanes-algorithm-dynamic-programming-how-and-why-does-it-work-3fd8849ed73d
+    Kadane's algorithm to get maximum subarray sum
+    Please enter the decimal values only and use spaces as separator
+"""
+def negative_exist(arr):
+
+    """
+      >>> negative_exist([-2,-8,-9])
+      -2
```
Comment (cclauss) [inline maths/kadanes_algorithm.py:11]:
```suggestion
      -2
       >>> [negative_exist(arr) for arr in test_data]
       [-2, 0, 0, 0, 0]
```

File: maths/kadanes_algorithm.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+     arr = [2,3,-9,8,-2]
+       initially the value of max_sum = 0 and max_till_element is 0
+       than when max_sum less than max_till particular element it will assign that value to max_sum
+       and when value of max_till_sum is less than 0 it will assign 0 to i
+       and after whole process it will return the value
+
+       So the output for above arr is :- 8
+
+       >>> kadanes([2,3,-9,8,-2])
+       8
```
Comment (cclauss) [inline maths/kadanes_algorithm.py:39]:
```suggestion
       8
       >>> [kadanes(arr) for arr in test_data]
       [?, ?, ?, ?, ?]
```

**Comment by cclauss** [comment]
Nice contribution.  Thanks for doing this!

**Comment by Raj-Parekh24** [comment]
Thanks for the help.

And I am happy to contribute to your repository.

**Comment by cclauss** [comment]
> I am happy to contribute to your repository.

It is not my repository, it is _our_ repository.  Welcome to the club.  ;-)

**Comment by cclauss** [comment]
We have a problem.  This implementation and the first `max_subarray()` implementation at https://en.wikipedia.org/wiki/Maximum_subarray_problem#Kadane's_algorithm disagree about the answer for `[-2, -8, -9]`.

`kadanes([-2, -8, -9])`  # -2
`max_subarray([-2, -8, -9])`  # `0`

Please investigate.

Note: the _second_ implementation on the Wikipedia page crashes when trying to run our tests.  :-(

**Comment by Raj-Parekh24** [comment]
sorry for the late reply due to examination i did not check my mails.
According to kadanes it will give a max -ve number and for max_sub array it
will return 0.
And max_sub array is already implemented in the repository.

**Pull request merged**